File: backend/app/models/rooms.py
"""Hebergement : etages, types de chambres, chambres, tarifs, taxes."""

from __future__ import annotations

import datetime as dt
import uuid

from sqlalchemy import (
    BigInteger,
    Boolean,
    CheckConstraint,
    Date,
    Enum,
    ForeignKey,
    Index,
    Integer,
    String,
    Text,
    UniqueConstraint,
)
from sqlalchemy.dialects.postgresql import JSONB
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.db.base import HotelScoped, RefBase, SyncBase
from app.models.enums import HousekeepingStatus, OccupancyStatus, TaxMode


def _enum(e, name: str):
    """VARCHAR + CHECK plutot qu'un ENUM natif (voir app/models/enums.py)."""
    return Enum(e, name=name, native_enum=False, length=32, validate_strings=True)


class Floor(RefBase, HotelScoped):
    __tablename__ = "floors"
    __table_args__ = (UniqueConstraint("hotel_id", "code"),)

    code: Mapped[str] = mapped_column(String(16))
    label: Mapped[str] = mapped_column(String(80))
    sort_order: Mapped[int] = mapped_column(default=0)
    # Fond de plan de l'etage pour l'ecran graphique du paragraphe 5.2.
    map_image_path: Mapped[str | None] = mapped_column(String(255), default=None)


class RoomType(RefBase, HotelScoped):
    __tablename__ = "room_types"
    __table_args__ = (UniqueConstraint("hotel_id", "code"),)

    code: Mapped[str] = mapped_column(String(16))
    label: Mapped[str] = mapped_column(String(80))
    description: Mapped[str | None] = mapped_column(Text, default=None)
    base_capacity: Mapped[int] = mapped_column(default=2)
    max_capacity: Mapped[int] = mapped_column(default=2)
    default_rate: Mapped[int] = mapped_column(BigInteger, default=0)
    amenities: Mapped[dict | None] = mapped_column(JSONB, default=None)
    photo_path: Mapped[str | None] = mapped_column(String(255), default=None)
    sort_order: Mapped[int] = mapped_column(default=0)


class Room(SyncBase, HotelScoped):
    """Chambre physique.

    Point de modelisation important : le cahier des charges (paragraphe 5.2)
    affiche un etat unique par pastille de couleur -- disponible, occupee,
    reservee, nettoyage, maintenance. C'est le bon affichage mais un mauvais
    stockage, car ces valeurs ne sont pas exclusives : une chambre peut etre
    occupee *et* en cours de nettoyage, ou en maintenance *et* sale.

    On conserve donc trois axes independants, et la pastille est calculee par
    `display_status`. Sans cela, la reception et le housekeeping s'ecrasent
    mutuellement a chaque synchronisation, puisqu'ils ecrivent le meme champ
    pour deux raisons differentes.

    Ces colonnes sont materialisees et non derivees des reservations : la
    tablette doit pouvoir peindre le plan de l'hotel instantanement et hors
    ligne, sans recalculer des intersections de dates sur toute la table.
    """

    __tablename__ = "rooms"
    __table_args__ = (
        UniqueConstraint("hotel_id", "number"),
        Index("ix_rooms_occupancy_status", "occupancy_status"),
        Index("ix_rooms_housekeeping_status", "housekeeping_status"),
    )

    number: Mapped[str] = mapped_column(String(16))
    room_type_id: Mapped[uuid.UUID] = mapped_column(
        ForeignKey("room_types.id", ondelete="RESTRICT"), index=True
    )
    floor_id: Mapped[uuid.UUID | None] = mapped_column(
        ForeignKey("floors.id", ondelete="SET NULL"), default=None, index=True
    )

    occupancy_status: Mapped[OccupancyStatus] = mapped_column(
        _enum(OccupancyStatus, "occupancy_status"), default=OccupancyStatus.VACANT
    )
    housekeeping_status: Mapped[HousekeepingStatus] = mapped_column(
        _enum(HousekeepingStatus, "housekeeping_status"),
        default=HousekeepingStatus.CLEAN,
    )
    is_out_of_order: Mapped[bool] = mapped_column(Boolean, default=False)
    out_of_order_reason: Mapped[str | None] = mapped_column(String(255), default=None)
    out_of_order_until: Mapped[dt.date | None] = mapped_column(Date, default=None)

    # Coordonnees sur le plan interactif (F1.7).
    map_x: Mapped[int | None] = mapped_column(default=None)
    map_y: Mapped[int | None] = mapped_column(default=None)

    phone_ext: Mapped[str | None] = mapped_column(String(16), default=None)
    notes: Mapped[str | None] = mapped_column(Text, default=None)
    is_active: Mapped[bool] = mapped_column(Boolean, default=True)

    room_type: Mapped[RoomType] = relationship(lazy="joined")
    floor: Mapped[Floor | None] = relationship(lazy="joined")

    @property
    def display_status(self) -> str:
        """Pastille de l'ecran 5.2, par ordre de priorite decroissante."""
        if self.is_out_of_order:
            return "MAINTENANCE"
        if (
            self.housekeeping_status
            in (
                HousekeepingStatus.IN_PROGRESS,
                HousekeepingStatus.DIRTY,
            )
            and self.occupancy_status is OccupancyStatus.VACANT
        ):
            return "CLEANING"
        if self.occupancy_status is OccupancyStatus.OCCUPIED:
            return "OCCUPIED"
        if self.occupancy_status is OccupancyStatus.RESERVED:
            return "RESERVED"
        return "AVAILABLE"


class RatePlan(RefBase, HotelScoped):
    __tablename__ = "rate_plans"
    __table_args__ = (UniqueConstraint("hotel_id", "code"),)

    code: Mapped[str] = mapped_column(String(32))
    label: Mapped[str] = mapped_column(String(120))
    room_type_id: Mapped[uuid.UUID | None] = mapped_column(
        ForeignKey("room_types.id", ondelete="CASCADE"), default=None, index=True
    )
    is_default: Mapped[bool] = mapped_column(Boolean, default=False)
    valid_from: Mapped[dt.date | None] = mapped_column(Date, default=None)
    valid_to: Mapped[dt.date | None] = mapped_column(Date, default=None)
    min_nights: Mapped[int] = mapped_column(default=1)
    includes_breakfast: Mapped[bool] = mapped_column(Boolean, default=False)
    description: Mapped[str | None] = mapped_column(Text, default=None)


class RatePlanPrice(SyncBase):
    """Prix d'un plan tarifaire sur une periode et certains jours de semaine."""

    __tablename__ = "rate_plan_prices"
    __table_args__ = (
        Index("ix_rate_plan_prices_range", "rate_plan_id", "date_from", "date_to"),
        CheckConstraint("date_to >= date_from", name="date_range"),
    )

    rate_plan_id: Mapped[uuid.UUID] = mapped_column(
        ForeignKey("rate_plans.id", ondelete="CASCADE")
    )
    date_from: Mapped[dt.date] = mapped_column(Date)
    date_to: Mapped[dt.date] = mapped_column(Date)
    # Masque binaire des jours concernes : bit 0 = lundi ... bit 6 = dimanche.
    # 127 = toute la semaine. Permet un tarif week-end sans dupliquer les lignes.
    weekday_mask: Mapped[int] = mapped_column(default=127)
    price: Mapped[int] = mapped_column(BigInteger)


class Tax(RefBase, HotelScoped):
    """Taxe applicable (TVA, taxe de sejour, taxe de promotion touristique)."""

    __tablename__ = "taxes"
    __table_args__ = (UniqueConstraint("hotel_id", "code"),)

    code: Mapped[str] = mapped_column(String(32))
    label: Mapped[str] = mapped_column(String(120))
    mode: Mapped[TaxMode] = mapped_column(_enum(TaxMode, "tax_mode"))
    # Taux en pourcentage, ou montant fixe selon `mode`.
    rate: Mapped[int] = mapped_column(Integer)
    # Categories de charges concernees, par exemple ["ROOM", "FNB"].
    applies_to: Mapped[dict | None] = mapped_column(JSONB, default=None)
    # Taxe deja comprise dans le prix affiche (TTC) ou ajoutee au moment de la
    # facturation (HT). Change entierement le calcul de la facture.
    is_included: Mapped[bool] = mapped_column(Boolean, default=True)
    sort_order: Mapped[int] = mapped_column(default=0)

File: frontend/lib/data/local/tables/rooms.dart
/// Hebergement : etages, types de chambres, chambres, tarifs, taxes.
library;

import 'package:drift/drift.dart';

import '../columns.dart';
import '../enums.dart';

@DataClassName('FloorRow')
class Floors extends Table with SyncedTableColumns, HotelScoped, RefTableColumns {
  TextColumn get code => text().withLength(max: 16)();
  TextColumn get label => text().withLength(max: 80)();
  IntColumn get sortOrder => integer().withDefault(const Constant(0))();

  /// Fond de plan de l'etage, pour l'ecran graphique du paragraphe 5.2.
  TextColumn get mapImagePath => text().withLength(max: 255).nullable()();
}

@DataClassName('RoomTypeRow')
class RoomTypes extends Table
    with SyncedTableColumns, HotelScoped, RefTableColumns {
  TextColumn get code => text().withLength(max: 16)();
  TextColumn get label => text().withLength(max: 80)();
  TextColumn get description => text().nullable()();
  IntColumn get baseCapacity => integer().withDefault(const Constant(2))();
  IntColumn get maxCapacity => integer().withDefault(const Constant(2))();

  /// Tarif de reference, en francs CFA, entiers.
  IntColumn get defaultRate => integer().withDefault(const Constant(0))();

  /// Equipements, en JSON. Chaque etablissement suit sa propre liste et la
  /// fait evoluer : un schema fige imposerait une migration a chaque ajout.
  TextColumn get amenities => text().nullable()();
  TextColumn get photoPath => text().withLength(max: 255).nullable()();
  IntColumn get sortOrder => integer().withDefault(const Constant(0))();
}

/// Chambre physique.
///
/// Le point de modelisation le plus important du module. Le cahier des charges
/// (paragraphe 5.2) affiche un etat unique par pastille de couleur :
/// disponible, occupee, reservee, nettoyage, maintenance. C'est le bon
/// affichage, mais ce serait un mauvais stockage, parce que ces valeurs ne
/// s'excluent pas : une chambre peut etre occupee *et* en cours de nettoyage,
/// ou en maintenance *et* sale.
///
/// On conserve donc trois axes independants, et la pastille est calculee a
/// l'affichage. Si l'on n'avait qu'une colonne, la reception et le
/// housekeeping ecriraient la meme case pour deux raisons differentes et
/// s'ecraseraient mutuellement a chaque synchronisation -- exactement le genre
/// de perte de donnee que le mode hors ligne rend inevitable.
///
/// Ces colonnes sont materialisees plutot que deduites des reservations : le
/// plan de l'hotel doit s'afficher instantanement et hors ligne, sans croiser
/// des intervalles de dates sur toute la table des sejours.
@DataClassName('RoomRow')
class Rooms extends Table with SyncedTableColumns, HotelScoped {
  TextColumn get number => text().withLength(max: 16)();
  TextColumn get roomTypeId => text().withLength(min: 36, max: 36)();
  TextColumn get floorId => text().nullable()();

  TextColumn get occupancyStatus => textEnum<OccupancyStatus>()
      .withDefault(const Constant('VACANT'))();
  TextColumn get housekeepingStatus => textEnum<HousekeepingStatus>()
      .withDefault(const Constant('CLEAN'))();

  BoolColumn get isOutOfOrder => boolean().withDefault(const Constant(false))();
  TextColumn get outOfOrderReason => text().withLength(max: 255).nullable()();

  /// Date au format AAAA-MM-JJ.
  TextColumn get outOfOrderUntil => text().withLength(max: 10).nullable()();

  /// Position sur le plan interactif (F1.7).
  IntColumn get mapX => integer().nullable()();
  IntColumn get mapY => integer().nullable()();

  TextColumn get phoneExt => text().withLength(max: 16).nullable()();
  TextColumn get notes => text().nullable()();
  BoolColumn get isActive => boolean().withDefault(const Constant(true))();
}

@DataClassName('RatePlanRow')
class RatePlans extends Table
    with SyncedTableColumns, HotelScoped, RefTableColumns {
  TextColumn get code => text().withLength(max: 32)();
  TextColumn get label => text().withLength(max: 120)();
  TextColumn get roomTypeId => text().nullable()();
  BoolColumn get isDefault => boolean().withDefault(const Constant(false))();
  TextColumn get validFrom => text().withLength(max: 10).nullable()();
  TextColumn get validTo => text().withLength(max: 10).nullable()();
  IntColumn get minNights => integer().withDefault(const Constant(1))();
  BoolColumn get includesBreakfast =>
      boolean().withDefault(const Constant(false))();
  TextColumn get description => text().nullable()();
}

@DataClassName('RatePlanPriceRow')
class RatePlanPrices extends Table with SyncedTableColumns {
  TextColumn get ratePlanId =>
      text().references(RatePlans, #id, onDelete: KeyAction.cascade)();
  TextColumn get dateFrom => text().withLength(max: 10)();
  TextColumn get dateTo => text().withLength(max: 10)();

  /// Masque binaire des jours concernes : bit 0 = lundi, bit 6 = dimanche.
  /// 127 couvre toute la semaine. Permet un tarif de week-end distinct sans
  /// dupliquer une ligne par date.
  IntColumn get weekdayMask => integer().withDefault(const Constant(127))();

  /// Prix en francs CFA, entiers.
  IntColumn get price => integer().withDefault(const Constant(0))();
}

@DataClassName('TaxRow')
class Taxes extends Table with SyncedTableColumns, HotelScoped, RefTableColumns {
  TextColumn get code => text().withLength(max: 32)();
  TextColumn get label => text().withLength(max: 120)();
  TextColumn get mode => textEnum<TaxMode>()();

  /// Taux en points de base pour le mode PERCENT (18 % vaut 1800), montant en
  /// francs CFA entiers pour les modes forfaitaires.
  IntColumn get rate => integer().withDefault(const Constant(0))();

  /// Categories de charges concernees, en JSON.
  TextColumn get appliesTo => text().nullable()();

  /// Taxe comprise dans le prix affiche, ou ajoutee au moment de la
  /// facturation. Ce drapeau change entierement le calcul de la facture, il ne
  /// peut pas etre une convention implicite.
  BoolColumn get isIncluded => boolean().withDefault(const Constant(true))();
  IntColumn get sortOrder => integer().withDefault(const Constant(0))();
}
